### PR TITLE
Issue 1911 added search query optimization parameters for Postgresql

### PR DESCRIPTION
**Pull request by punktilious**

### CODE REVIEW

File: fhir-config/src/main/java/com/ibm/fhir/config/FHIRConfiguration.java [hunk 1]
```diff
@@ -105,6 +105,9 @@
     public static final String PROPERTY_JDBC_EXTERNAL_REF_SYSTEM_CACHE_SIZE = "fhirServer/persistence/jdbc/externalRefSystemCacheSize";
     public static final String PROPERTY_JDBC_EXTERNAL_REF_VALUE_CACHE_SIZE = "fhirServer/persistence/jdbc/externalRefValueCacheSize";
 
+    // Optimizer options within a datasource definition
```
Comment (prb112) [inline fhir-config/src/main/java/com/ibm/fhir/config/FHIRConfiguration.java:108]: Copyright Year needs to be bumped to 2021

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/connection/FHIRDbConnectionStrategy.java [hunk 2]
```diff
@@ -16,7 +16,7 @@
  * Abstraction used to obtain JDBC connections. The database being connected
  * is determined by the datasource currently referenced by the {@link FHIRRequestContext}
  * (which is a ThreadLocal thing, and not obvious).
- * 
+ *
```
Comment (prb112) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/connection/FHIRDbConnectionStrategy.java:19]: Copyright Year to 2021

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/connection/FHIRDbConnectionStrategyBase.java [hunk 3]
```diff
@@ -21,6 +21,7 @@
 import com.ibm.fhir.database.utils.model.DbType;
 import com.ibm.fhir.persistence.exception.FHIRPersistenceException;
 import com.ibm.fhir.persistence.jdbc.exception.FHIRPersistenceDataAccessException;
+import com.ibm.fhir.persistence.jdbc.postgresql.SetPostgresOptimizerOptions;
```
Comment (prb112) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/connection/FHIRDbConnectionStrategyBase.java:24]: Copyright Year to 2021

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/connection/FHIRDbTestConnectionStrategy.java [hunk 4]
```diff
@@ -16,32 +16,32 @@
 
 /**
  * Hides the logic behind obtaining a JDBC {@link Connection} from the DAO code.
- * 
+ *
```
Comment (prb112) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/connection/FHIRDbTestConnectionStrategy.java:19]: Copyright Year to 2021

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java [hunk 5]
```diff
@@ -327,6 +327,7 @@ protected Action buildActionChain() {
         // reads/searches.
         result = new CreateTempTablesAction(result);
 
+        // For PostgreSQL
```
Comment (prb112) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java:330]: Copyright Year to 2021

File: fhir-server/liberty-config/config/default/fhir-server-config-postgresql.json [hunk 6]
```diff
@@ -59,7 +59,11 @@
             "datasources": {
                 "default": {
                     "type": "postgresql",
-                    "currentSchema": "fhirdata"
+                    "currentSchema": "fhirdata",
+                    "searchOptimizerOptions": {
+                        "from_collapse_limit": 16,
+                        "join_collapse_limit": 16
```
Comment (prb112) [inline fhir-server/liberty-config/config/default/fhir-server-config-postgresql.json:65]: Needs documentation on the new feature in the FHIRServerUsersGuide

**Review by prb112 — COMMENTED** [review]
Looks good, needs some documentation

File: fhir-server/liberty-config/config/default/fhir-server-config-postgresql.json (hunk 6 above)
Comment (punktilious) [reply fhir-server/liberty-config/config/default/fhir-server-config-postgresql.json:65]: added

**Review by lmsurpre — APPROVED** [review]
LGTM

**Review by michaelwschroeder — APPROVED** [review]
LGTM - just a couple comments on the doc

File: fhir-config/src/main/java/com/ibm/fhir/config/FHIRConfiguration.java (hunk 1 above)
Comment (punktilious) [reply fhir-config/src/main/java/com/ibm/fhir/config/FHIRConfiguration.java:108]: fixed

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/connection/FHIRDbConnectionStrategyBase.java (hunk 3 above)
Comment (punktilious) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/connection/FHIRDbConnectionStrategyBase.java:24]: fixed

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/connection/FHIRDbTestConnectionStrategy.java (hunk 4 above)
Comment (punktilious) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/connection/FHIRDbTestConnectionStrategy.java:19]: fixed

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java (hunk 5 above)
Comment (punktilious) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java:330]: fixed

File: docs/src/pages/guides/FHIRServerUsersGuide.md [hunk 7]
```diff
@@ -2221,6 +2221,8 @@ This section contains reference information about each of the configuration prop
...
 |`fhirServer/bulkdata/useFhirServerTrustStore`|boolean| If the COS Client should use the IBM FHIR Server's TrustStore to access S3/IBMCOS service |
 |`fhirServer/bulkdata/enableParquet`|boolean| Whether or not the server is configured to support export to parquet; to properly enable it the administrator must first make spark and stocator available to the fhir-bulkimportexport-webapp (e.g through the shared lib at `wlp/user/shared/resources/lib`) |
+|`fhirServer/persistence/datasources/<datasourceId>/searchOptimizerOptions/from_collapse_limit`|int| For PostgreSQL, sets the from_collapse_limit query optimizer parameter to improve search performance. If not set, the IBM FHIR Server uses a value of 16. To use the database default (8), explicitly set this value to null. |
+|`fhirServer/persistence/datasources/<datasourceId>/searchOptimizerOptions/join_collapse_limit`|int| For PostgreSQL, sets the join_collapse_limit query optimizer parameter to improve search performance. If not set, the IBM FHIR Server uses a value of 16. To use the database default (8), explicitly set this value to null. |
```
[the remark this answers is not in the document]
Comment (punktilious) [reply docs/src/pages/guides/FHIRServerUsersGuide.md:2225]: ok

File: docs/src/pages/guides/FHIRServerUsersGuide.md [hunk 8]
```diff
@@ -2270,6 +2272,8 @@ This section contains reference information about each of the configuration prop
 |`fhirServer/persistence/datasources`|embedded Derby database: derby/fhirDB|
 |`fhirServer/persistence/jdbc/dataSourceJndiName`|jdbc/fhirProxyDataSource|
 |`fhirServer/persistence/jdbc/bootstrapDb`|false|
+|`fhirServer/persistence/datasources/<datasourceId>/searchOptimizerOptions/from_collapse_limit`|16|
+|`fhirServer/persistence/datasources/<datasourceId>/searchOptimizerOptions/join_collapse_limit`|16|
```
[the remark this answers is not in the document]
Comment (punktilious) [reply docs/src/pages/guides/FHIRServerUsersGuide.md:2276]: done

**Pull request merged**